src/jit/hwintrinsicxarch.cpp

```diff
@@ -295,6 +301,125 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
+// impNonConstFallback: convert certain SSE2/AVX2 shift intrinsic to its semantic alternative when the imm-arg is
+// not a compile-time constant
+//
+// Arguments:
+//    intrinsic  -- intrinsic ID
+//    simdType   -- Vector type
+//    baseType   -- base type of the Vector128/256<T>
+//
+// Return Value:
+//     return the IR of semantic alternative on non-const imm-arg
+//
+GenTree* Compiler::impNonConstFallback(NamedIntrinsic intrinsic, var_types simdType, var_types baseType)
+{
+    assert((flagsOfHWIntrinsic(intrinsic) & HW_Flag_NoJmpTableIMM) != 0);
+    switch (intrinsic)
+    {
+        case NI_SSE2_ShiftLeftLogical:
+        case NI_SSE2_ShiftRightArithmetic:
+        case NI_SSE2_ShiftRightLogical:
+        case NI_AVX2_ShiftLeftLogical:
+        case NI_AVX2_ShiftRightArithmetic:
+        case NI_AVX2_ShiftRightLogical:
+        {
+            GenTree* op2 = impPopStack().val;
+            GenTree* op1 = impSIMDPopStack(simdType);
+            GenTree* tmpOp =
+                gtNewSimdHWIntrinsicNode(TYP_SIMD16, op2, NI_SSE2_ConvertScalarToVector128Int32, TYP_INT, 16);
+            return gtNewSimdHWIntrinsicNode(simdType, op1, tmpOp, intrinsic, baseType, genTypeSize(simdType));
+        }
+
+        default:
+            unreached();
+            return nullptr;
+    }
+}
+
+//------------------------------------------------------------------------
+// isImmHWIntrinsic: check the intrinsic is a imm-intrinsic overload or not
+//
+// Arguments:
+//    intrinsic  -- intrinsic ID
+//    lastOp     -- the last operand of the intrinsic that may point to the imm-arg
+//
+// Return Value:
+//     return if the intrinsic is a imm-intrinsic overload or not because some IMM-intrinsic (HW_Flag_MaybeIMM) have
+//     both imm and vector overloads
+//
+bool Compiler::isImmHWIntrinsic(NamedIntrinsic intrinsic, GenTree* lastOp)
+{
+    if (categoryOfHWIntrinsic(intrinsic) != HW_Category_IMM)
+    {
+        return false;
+    }
+
+    if ((flagsOfHWIntrinsic(intrinsic) & HW_Flag_MaybeIMM) != 0 && genActualType(lastOp->TypeGet()) != TYP_INT)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+//------------------------------------------------------------------------
+// addRangeCheck: add a GT_HW_INTRINSIC_CHK node for non-full-range imm-intrinsic
+//
+// Arguments:
+//    intrinsic  -- intrinsic ID
+//    lastOp     -- the last operand of the intrinsic that points to the imm-arg
+//    mustExpand -- true if the compiler is compiling the fallback(GT_CALL) of this intrinsics
+//
+// Return Value:
+//     add a GT_HW_INTRINSIC_CHK node for non-full-range imm-intrinsic, which would throw ArgumentOutOfRangeException
+//     when the imm-argument is not in the valid range
+//
+GenTree* Compiler::addRangeCheck(NamedIntrinsic intrinsic, GenTree* lastOp, bool mustExpand)
+{
+    assert(lastOp != nullptr);
+    if (mustExpand && ((flagsOfHWIntrinsic(intrinsic) & HW_Flag_FullRangeIMM) == 0) &&
```

Comment:
It would be useful to add a comment explaining why "full range imm" intrinsics do not need a range check - because the imm parameter of the intrinsic method is a `byte`.

That BTW might be a bit unfortunate, at least for shift intrinsics. Both the C/C++ intrinsics and the C/C++/C# shift operators use `int` or `unsigned` for the shift count. Using `byte` can be considered as a rather arbitrary restriction but probably it doesn't matter as we expect people to supply constants.

Reply:
> It would be useful to add a comment explaining why "full range imm" intrinsics do not need a range check

Will do.